torch/csrc/Exceptions.h

```diff
@@ -366,30 +366,37 @@ struct PyWarningHandler {
 };
 
 namespace detail {
+
+struct noop_gil_scoped_release {
+  noop_gil_scoped_release() {
+    // suppress `unused variable` error messages at call sites
+    (void)(this != (this + 1));
```

Comment:
Can the constructor be made default and remove the strange this comparison?

Reply:
No, see the comment. I basically copied that from pybind11: https://github.com/pybind/pybind11/blob/f29def9ea467c7fde754440733047953fa4b990c/include/pybind11/gil.h#L226-L227

Background: If the constructor would be trivial/default every `noop_gil_scoped_release foo;` would trigger an unused variable warning.   
This code clearly does nothing and will be optimized away and it's purpose is documented. So IMO this is much better than the alternative of using an unused-attribute at each use site.

Reply:
Would a C10_UNUSED; be enough here?
If not, I think we should just define a C10_NOOP otherwise other people are going to get confused by this code.

Reply:
No because the warning occurs at the usage site. So you'd need `C10_UNUSED noop_gil_scoped_release foo;` at all sites potentially depending on what is actually used.

Reply:
I found that simply defining the ctor is enough, even when it is empty. I want to avoid the `NOOP` macro as this place is literally the only use case for it.   
If more classes would need this they would all look exactly like this, so this one could be renamed to something like `noop_scope_guard`

I changed the ctor and comment to be clearer